rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+
+```
+// A simple toy module called "M"
+typedef struct M_context;
+
+// The summary of the interface plugins need to implement
+typedef struct MInterface {
+  void (*M_Init)(M_context*, string);
+  void (*M_Execute)(M_context*);
+};
+
+// The plugin registry for module M
+// The registry here implemented uses C++ class std::map
+// This is OK as it is not exposed in our example here.
+// As far as module implementations are concerned, they only
+// need to see and use M_RegisterPlugin method, and that method takes // care of their addition into this registry.
+std::map<string, MInterface> m_plugin_registry;
+
+// Function to call for each plugin at load time
+// to add themselves to the registry
+void M_RegisterPlugin(
+    string plugin, 
+    void (*M_Init)(M_context*, string),
+    void (*M_Execute)(M_context*)) {
+  // Omitting error handling.
+  m_plugin_registry.insert(plugin,
+                           MInterface(M_Init, M_Execute));
+}
+
+// Implementation of the interface is just a thin layer to
+// get the correct plugin and call it.
+// Here we assume that plugin explicitly gets selected by
+// the init function. Some modules can go with implicit selection,
+// Such as deducing the filesystem from the file path.
+void M_Init(M_context* ctx, string id) {
+  // Just a quick hacky way to select the plugin here
+  ctx->interface = m_plugin_registry[id];
+  ctx->interface.M_Init(ctx, id);
+}
+
+void M_Execute(M_context* ctx) {
+  ctx->interface.M_Execute(ctx);
+}
+```
+
+
+**Please note that the above is a toy example to explain the abstract ideas described above. Exact implementation can vary across different modules.**
+
+Interface has to be pure C at the ABI level. We can have C++ header-only libraries built on top of these C ABI/APIs.
+
+
+#### Plugins
+
+Plugins need to include implementation of the interfaces declared by one module. If the module interface requires Init and Compute methods, it will need to implement these two functions, plus a TF_InitPlugin function which will be called at load time. This function will also need to register the plugin as prescribed by the module.
+
+
+```
+// A simple plugin implementation A, for module M.
+typedef struct M_context;
+
+// Here is the meat of the plugin:
+void A_Init(M_context* ctx, string id) {
+  // id can be thrown away, in this example.
+  // Or can encode different ways plugin can be initialized.
+  ......
+  // initialize the plugin
+  // Modify the context using functions exposed by core
+  TF_SetFooInContext(ctx, foo);
+  TF_InitBarInContext(ctx);
+}
+
+void A_Execute(M_context* ctx) {
+  ..........
+}
+
+void TF_InitPlugin() {
+  M_RegisterPlugin("A", A_Init, A_Execute);
+}
+```
+
+
+When this plugin is loaded by TF at runtime, `TF_InitPlugin` method will be called. This method will register the plugin as prescribed by the module, and exit.
```

Comment:
How will conflicts and priority be decided between plugins (e.g. CPU and GPU)?

Reply:
We will try to avoid conflicts.
For ops, we have this rough idea: Each op plugin is required to use a separate "namespace" assigned to them for creating new ops. But we will have to improve on this idea.

Reply:
Clarification: if two plugins register the ability to handle the same op (or whatever granularity of operation), how will priority be decided between them?

Reply:
In the current implementation of the kernel registry, you can add extra labels to each kernel being registered, and use those to select which kernels you would like to pick:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/kernel_def.proto#L35
Our initial idea is to leverage this, but @sjamesr may have other ideas.